service/src/main/java/bio/terra/externalcreds/logging/LoggerInterceptor.java

```diff
@@ -0,0 +1,127 @@
+
+    String userId;
+    String userEmail;
+    try {
+      AuthenticatedUserRequest userReq = authenticatedUserRequestFactory.from(request);
+      userId = userReq.getSubjectId();
+      userEmail = userReq.getEmail();
+    } catch (UnauthorizedException e) {
+      userId = UNAUTHED_USER_ID;
+      userEmail = UNAUTHED_EMAIL;
+    }
+
+    String url = request.getServletPath();
+    String method = request.getMethod();
+    Map<String, String[]> paramMap = request.getParameterMap();
+    Gson gson = new Gson();
+    String paramString = gson.toJson(paramMap);
+    String responseStatus = Integer.toString(response.getStatus());
+    long requestDuration;
+    Long requestStartTime = (Long) request.getAttribute(REQUEST_START_ATTRIBUTE);
+    if (requestStartTime != null) {
+      requestDuration = System.currentTimeMillis() - requestStartTime;
+    } else {
+      requestDuration = NOT_FOUND_DURATION;
+    }
+    // skip logging the status endpoint
+    if (LOG_EXCLUDE_LIST.stream().noneMatch(url::endsWith)) {
+
+      String requestPath;
+      try {
+        URI uri = new URI(request.getRequestURI());
+        requestPath = uri.getPath();
+      } catch (URISyntaxException e) {
+        log.error("Error parsing request path. Logging the full URI instead.", e);
+        requestPath = request.getRequestURI();
+      }
+
+      Map<String, String> stackDriverPayload = new HashMap<>();
+      if (RequestMethod.POST.name().equalsIgnoreCase(method)
+          || RequestMethod.PUT.name().equalsIgnoreCase(method)) {
+        String requestBody =
+            new String(
+                ((ContentCachingRequestWrapper) request).getContentAsByteArray(),
+                StandardCharsets.UTF_8);
+        String requestBodyPayload =
+            requestBody.length() > STACKDRIVER_MAX_CHARS
+                ? requestBody.substring(0, STACKDRIVER_MAX_CHARS)
+                : requestBody;
+        stackDriverPayload.put("requestBody", requestBodyPayload);
+        stackDriverPayload.put("userId", userId);
+        stackDriverPayload.put("userEmail", userEmail);
+        stackDriverPayload.put("params", paramString);
+        stackDriverPayload.put("duration", Long.toString(requestDuration));
+      }
+      // Log the message, and include the supplementary JSON as an additional arg.
+      // If GoogleJsonLayout has been loaded, it will merge the JSON into the structured log output
+      // for ingestion by Cloud Logging. If the default logback layout is being used, the JSON
+      // argument will be ignored.
+      String message = String.format("%s %s %s", method, requestPath, responseStatus);
+      if (response.getStatus() >= 400) {
+        log.warn(message, stackDriverPayload);
+      } else {
+        log.info(message, stackDriverPayload);
+      }
+    } else {
+      log.debug("Received request at {}", url);
+    }
+
+    if (ex != null) {
+      log.error("An error occurred processing this request: ", ex);
```

Comment:
would it be helpful to also include `stackDriverPayload` for exceptions? Do we already get this for "free" since an exception in our code would result in a 500 error (which would be logged on L115)?

Reply:
We could return 4xx responses for reasons that don't throw an exception. If a user's passport doesn't validate, we shouldn't throw an exception, but still return 400 to the user. Also, hiding exceptions in in the stackdriver payload in our logs makes them less discoverable, which is something we don't want.

Reply:
I was asking my questions to see if it would help the traceability and debug-ability of exceptions, but it sounds like it might do the opposite, in which case I'm with you